subsys/instrumentation/common/instr_common.c

```diff
@@ -0,0 +1,561 @@
+#include <zephyr/kernel_structs.h>
+#include <zephyr/retention/retention.h>
+#include <zephyr/sys/reboot.h>
+
+#include <kernel_internal.h>
+#include <ksched.h>
+
+/*
+ * Memory buffer to store instrumentation event records has the following modes:
+ *
+ * Callgraph (tracing) ring buffer (default): Replace oldest entries when buffer
+ * is full.
+ *
+ * Callgraph (tracing) fixed buffer: Stop buffering events when the buffer is
+ * full, ensuring we have a callgraph from reset point or from wherever the
+ * trigger function was called for the first time.
+ *
+ * Statistical (profiling): Buffer functions until out of memory.
+ *
+ */
+
+const struct device *retention0 = DEVICE_DT_GET(DT_NODELABEL(retention0));
+
+static bool _instr_initialized;
+static bool _instr_enabled;
+static bool _instr_on;
+static bool _instr_tracing_disabled;
+static bool _instr_profiling_disabled;
+static bool _instr_tracing_supported = IS_ENABLED(CONFIG_INSTRUMENTATION_MODE_CALLGRAPH);
+static bool _instr_profiling_supported = IS_ENABLED(CONFIG_INSTRUMENTATION_MODE_STATISTICAL);
+
+#if defined(CONFIG_INSTRUMENTATION_MODE_STATISTICAL)
+/*
+ * Entry for discovered functions. The functions are added to 'disco_func' array
+ * as they are called in the execution flow, hence "discovered" functions. Once
+ * NUM_DISCO_FUNCS is reached, additional new executed functions are ignored and
+ * no profiling information is collected for them.
+ */
+
+#define MAX_CALL_DEPTH CONFIG_INSTRUMENTATION_MODE_STATISTICAL_MAX_CALL_DEPTH
+struct disco_func_entry {
+	timing_t entry_timestamp;		/* Timestamp at function entry */
+	uint64_t delta_t;			/* Accumulated (per function) delta time */
+	void *addr;				/* Function address/ID */
+	uint16_t call_depth;			/* Call depth */
+};
+
+#define MAX_NUM_DISCO_FUNC CONFIG_INSTRUMENTATION_MODE_STATISTICAL_MAX_NUM_FUNC
+static int num_disco_func;
+struct disco_func_entry disco_func[MAX_NUM_DISCO_FUNC] = { 0 };
+
+/* To track the number of unbalanced/spurious entry/exist pairs, for debugging */
+int unbalanced;
+#endif
+
+/* See instr_fundamentals_initialized() */
+uint16_t magic = 0xABBA;
+
+/* Defined in Kconfig */
+extern void INSTR_TRIGGER_FUNCTION(void);
+extern void INSTR_STOPPER_FUNCTION(void);
+
+/* Default trigger and stopper addresses, from Kconfig */
+void *k_trigger_callee = INSTR_TRIGGER_FUNCTION;
+void *k_stopper_callee = INSTR_STOPPER_FUNCTION;
+
+/* Current (live) trigger and stopper addresses */
+void *trigger_callee;
+void *stopper_callee;
+
+bool instr_tracing_supported(void)
+{
+	return _instr_tracing_supported;
+}
+
+bool instr_profiling_supported(void)
+{
+	return _instr_profiling_supported;
+}
+
+__no_instrumentation__
+int instr_init(void)
+{
+	/*
+	 * This flag needs to be set before calling any other function, otherwise it will cause an
+	 * infinite recursion in the handler since instr_initialized() will return 0 and
+	 * instr_init() will be called again.
+	 */
+	_instr_initialized = 1;
+
+	if (retention_is_valid(retention0)) {
```

Comment:
Might want an assert to ensure either magic header or checksum are enabled in the area, otherwise this will always say the data is valid